adapters/repos/db/inverted/row_reader.go

```diff
@@ -89,16 +92,27 @@ func (rr *RowReader) Read(ctx context.Context, readFn ReadFn) error {
 // equal is a special case, as we don't need to iterate, but just read a single
 // row
 func (rr *RowReader) equal(ctx context.Context, readFn ReadFn) error {
-	if err := ctx.Err(); err != nil {
+	v, err := rr.equalHelper(ctx)
+	if err != nil {
 		return err
 	}
 
-	v, err := rr.bucket.SetList(rr.value)
+	_, err = readFn(rr.value, rr.transformToBitmap(v), rr.operator)
+	return err
+}
+
+// notEqual is another special case, as it's the opposite of equal. So instead
+// of reading just one row, we read all but one row.
+func (rr *RowReader) notEqual(ctx context.Context, readFn ReadFn) error {
+	v, err := rr.equalHelper(ctx)
 	if err != nil {
 		return err
 	}
 
-	_, err = readFn(rr.value, v)
+	// Invert the Equal results for an efficient NotEqual
+	inverted := roaringset.NewInvertedBitmap(
```

Comment:
`NewBitmapPrefill(...)` called inside `NewInvertedBitmap(...)` is rather expensive to run. We may want to create bitmap once, and reuse it later on.

One way to do that, could be factory that holds base bitmap filled with ids up to current maxId.
Then whenever bitmap of maxId is requested, base bitmap is populated with missing ids (if previous maxId is lower than requested maxId), cloned and returned. Cloning should be cheap - it is just copying of bm's binary representation. Returned bm can then be used to compute inverted bm.
Base bitmap (the one to be cloned) could be populated on 1st request or on shard's startup. It can be initially populated (or expanded later on) to higher value than current maxId to reduce number of required expansions (e.g. init base bm with maxId + 1000; expand base bm to maxId + 1000 if bm's maxId is lower than requested one). 

wdyt?

Reply:
Sounds good, implemented [here](https://github.com/weaviate/weaviate/pull/4122/commits/9fb7bfdaf029d94807098b98f2d42f68e741471b)